### PR TITLE
Support whitespace in paths for ADD and COPY

### LINKED ISSUE #8318: COPY and ADD do not support filenames with whitespace.
This is because we use a whitespace-delimited parser there. We should probably support the JSON argument representation here. I don't think this will be a complicated code change.

/cc @shykes

### PR DESCRIPTION
**Pull request by icecrime**
This PR takes on @seongyeol37 work on supporting whitespace in paths for `ADD` and `COPY` Dockerfile instructions. It reverts an unrelated commit submitted as part of #8406, and adds tests and documentations for the new features.

Fixes #8318.
Replaces #8406.

### CODE REVIEW

**Comment by icecrime** [comment]
Ping @jfrazelle @LK4D4

**Comment by LK4D4** [comment]
yeah :) reviewing right now

**Comment by LK4D4** [comment]
Actually I don't like `[]` syntax requirement, there is no point in it. We have quote parsing code in `builder/parser/line_parsers.go:parseEnv`, which should be moved to separate function and used here.
ping @duglin, you're author of quote parser, maybe you want to take care of this?

**Comment by tianon** [comment]
Wait, we actually merged all that quote parsing code?  *facepalm*

**Comment by LK4D4** [comment]
@tianon haha :crying_cat_face:

**Comment by LK4D4** [comment]
Also, there is https://github.com/mattn/go-shellwords. mattn is pretty cool guy.

**Comment by icecrime** [comment]
See [@shykes comment](https://github.com/docker/docker/pull/8406#issuecomment-60465429) on previous PR. So @tianon @LK4D4 you both agree we should close this?

**Comment by tianon** [comment]
I'm the wrong person to ask -- I'm in the tiny minority who actually loves
how unambiguous and very verbosely standardized JSON is; I always know
exactly how it's going to parse, so I'm a big +1 on it being in Dockerfiles
and love that it's a thing.

**Comment by icecrime** [comment]
:sparkles: Summon `builder/` maintainers @erikh @tiborvass :sparkles:

**Comment by thaJeztah** [comment]
> I'm in the tiny minority who actually loves how unambiguous and very verbosely standardized JSON is

Perhaps not so tiny? Although I understand that the mixture of "Dockerfile" syntax, "shell" syntax and "JSON" in a *single* file may be confusing.

**Comment by LK4D4** [comment]
Json is cool, but we need and using mostly not json, but some cruel parody on json.

**Comment by tianon** [comment]
I wouldn't say parody, since what we use is 100% valid JSON -- it's just a
subset of the full flexibility of JSON, which isn't at all that unique to
see "in the wild".

**Comment by thaJeztah** [comment]
Perhaps Dockerfile V2 should include a JSON variant, ie, `Dockerfile.json`. Problem is that for *me*, writing JSON is my "bread and butter", but I'm mostly in web stuff, for others it may not be. Also rewriting shell (`RUN`) to JSON may be a bit troublesome or, at least, less readable,

But we're getting waaaaay off topic here (sorry).

**Comment by LK4D4** [comment]
@tianon No, you can write totally invalid json for `RUN` and it will work. Also try
```
CMD ["echo",  true]
```
and
```
CMD ["echo", 3]
```
and
```
CMD ["echo", {}]
```
pretty hard to predict what you get next from our parser, which works with standartized json.

**Comment by tianon** [comment]
I don't think we're really all that far off topic, since this is a PR that
implements more JSON.

Personally, I think the amount of JSON we have now is "just right", and
that forcing the entire file to be JSON is overkill.

The argument that JSON is hard for newbies or people who aren't familiar is
also pretty bogus because we get _far_ more issues with people not actually
understanding how the shell parses things than we do about JSON being
difficult.

**Comment by tianon** [comment]
_Subset_ of JSON.  ["a", "b", "c"] is valid JSON -- not at all
nonstandard.  It's also not even a little bit uncommon to expect very
specific types; in fact, Go's own JSON parser is based firmly on that very
idea (with struct tags enforcing it).

**Comment by thaJeztah** [comment]
> Personally, I think the amount of JSON we have now is "just right", and that forcing the entire file to be JSON is overkill.

Agreed, I was thinking that JSON would be easier to parse, validate and less room for "interpretation" mistakes. The problem right now is in the *maybe*JSON handling of arguments, because both "shell" syntax and JSON syntax is accepted and (as some other PRs pointed out) there is no way to determine if a parameter *is* JSON or just "looks like" JSON. Going full JSON would take away the guessing. Building an image through the API would also allow to send the Dockerfile as part of the JSON request;

```JSON
[
    {"instruction":"COPY", "params": ["/foo/bar", "/baz"]},
    {"instruction":"EXPOSE", "params": [80,443]}
]
```

But possibly, reading back today's discussion on `USER`, on [#docker-maintainers](https://botbot.me/freenode/docker-maintainers/2015-01-02/?msg=28657537&page=4), a "format" prefix could also work to prevent the guessing;

    user=foo format=JSON COPY ["............", "....."]

Where the above example would validate the argument for the `COPY` as JSON (not MaybeJSON)

**Comment by tianon** [comment]
Heh, something like this could be fun and simple:
```Dockerfile
RUN.json ["echo", "hi!"]
```

**Comment by thaJeztah** [comment]
Yes. But now that the `user=foo COPY ...` was decided on, I think adding `format=JSON` to that may be the most logical syntax to specify the format to use? (saw that you were discussing this issue right now on IRC)

**Comment by tianon** [comment]
Ah, I missed that.  I suppose format=... does make some sense then, but it
feels insanely verbose IMO.

**Comment by thaJeztah** [comment]
Yeah, but personally not 100% convinced on the `user=foo` at the start either, doesn't look as clean as `COPY` at position 0. (And how to handle a user with a space in its name?  ... More quoting magic?)

**Comment by duglin** [comment]
where is the user=foo stuff documented? I'd like to read more about that.

I think we added the quote support, initially, for ENV - not sure where else its used but I think the hope/plan is that in the v2 timeframe we switch to treat all commands like shell commands and then we can move away from this pseudo shell parsing we're doing now and do it for real.  And then also remove the JSON stuff since, as Shykes said, it was a stopgap thing. I don't think we should be adding more uses of it - especially if there isn't non-JSON equivalent support.

**Comment by thaJeztah** [comment]
@duglin the user=foo stuff was discussed and decided upon on IRC here: https://botbot.me/freenode/docker-maintainers/2015-01-02/?msg=28657537&page=4

**Comment by duglin** [comment]
wow.  Not thrilled with that one :-)
People are used to that being the syntax for setting env vars not passing in options to commands.  I would have much preferred if we just supported options on the commands. For example:
   COPY --foo=bar -p -q src dest
At least that's more what people are used to. This feels like the JSON parsing stopgap thing all over again, which we're trying to remove.  oh well.

**Comment by thaJeztah** [comment]
@duglin not very liking it either. Hope it will be opened up for discussion in a follow up PR, because (if not) I really wonder if the right procedure was followed for making such design decision. I mentioned this in a comment (https://github.com/docker/docker/pull/9046#issuecomment-68592752) on that PR, will see after the weekend if this becomes more clear.

**Comment by duglin** [comment]
Yea,its not clear to me whether that "decision" was to actually code it up or just to go off and do the "proposal" PR, which is just docs, so people can review that first - per the new process.

And totally on a tangent, but.... I have to admit, while I understand the desire for focused discussions, I find having to track (yet another) irc during the day is hard for me.  I didn't think docker-dev was that busy/noisy to warrant the split.  And at some point having too many things to follow will be perceived as "too hard" instead of "nicely open".  Especially, when the main one (now) appears to be read-only for most people.

But again, oh well.   :-)

**Comment by erikh** [comment]
hey @duglin -- honestly, the decision was the best of a bad lot in my opinion, at least. We would much prefer to modify the behavior of COPY.

Totally agree on the "too many channels" problem.

**Comment by icecrime** [comment]
Lots of truths in the previous comments, now to get back to this PR ;-) As far as I can tell:
1. We're not ditching JSON any time soon
2. @tianon PR#9881 nicely clarifies the subset of JSON allowed in Dockerfile

I'm leaving this open as I think it still makes sense to merge: now it's up to the maintainers ;p

**Comment by duglin** [comment]
@erikh was adding normal cmd line option parsing discussed?  And if so, why was it rejected?
I would think that:  
COPY -a src dest
and
COPY --opt=value src dest
doesn't really change the current behavior.  I _think_ the only worry is when one of the normal args might start with a dash but I can't imagine that's a huge concern. 
Although, to keep the fun going, we now get to discuss whether or not to support quotes in these new options that appear before the OP.   LOL

**Comment by erikh** [comment]
This  isn't the place for this; let's take this to irc tomorrow. Will be happy to answer questions. (fwiw, most of your questions are answered in that chat log above)

File: integration-cli/docker_cli_build_test.go [hunk 1]
```diff
@@ -762,41 +762,188 @@ RUN [ $(ls -l /exists/exists_file | awk '{print $3":"$4}') = 'dockerio:dockerio'
...
+			"test dir/test_file4": "test4",
+			"test_dir/test_file5": "test5",
+			"test dir/test_file6": "test6",
 		})
 	defer ctx.Close()
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	expected := "When using ADD with more than one source file, the destination must be a directory and end with a /"
```
Comment (erikh) [inline integration-cli/docker_cli_build_test.go]: why was this removed exactly?

**Comment by erikh** [comment]
LGTM presuming my comment is innocuous

File: integration-cli/docker_cli_build_test.go (hunk 1 above)
Comment (icecrime) [inline integration-cli/docker_cli_build_test.go]: I think the diff view is misleading: you can find it on line 784 of the patched file.

Comment (erikh) [inline integration-cli/docker_cli_build_test.go]: On Sun, Jan 04, 2015 at 03:06:28PM -0800, Arnaud Porterie wrote:
> >  		})
> >  	defer ctx.Close()
> >  	if err != nil {
> >  		t.Fatal(err)
> >  	}
> >  
> > -	expected := "When using ADD with more than one source file, the destination must be a directory and end with a /"
> 
> I think the diff view is misleading: you can find it on line 784 of the patched file.

Ah ok, no complaints, then!

-Erik

File: docs/man/Dockerfile.5.md [hunk 2]
```diff
@@ -131,7 +131,11 @@ or
  interactively, as with the following command: **docker run -t -i image bash**
 
 **ADD**
- --**ADD <src>... <dest>** The ADD instruction copies new files, directories
+ --ADD has two forms:
+ **ADD <src>... <dest>**
+ **ADD ["<src>"... "<dest>"]** This is the required form when paths contain
+ whitespaces.
```
Comment (fredlf) [inline docs/man/Dockerfile.5.md]: Suggest we simplify/clarify this to: "This form is required for paths containing whitespaces."

File: docs/sources/reference/builder.md [hunk 3]
```diff
@@ -381,7 +381,11 @@ change them using `docker run --env <key>=<value>`.
 
 ## ADD
 
-    ADD <src>... <dest>
+ADD has two forms:
+
+- `ADD <src>... <dest>`
+- `ADD ["<src>"... "<dest>"]` (the required form when path elements contain
+whitespace)
```
Comment (fredlf) [inline docs/sources/reference/builder.md]: For consistency, let's use the same phrase as above, please.

**Comment by fredlf** [comment]
Docs LGTM, with a couple of minor suggestions.

**Comment by icecrime** [comment]
Thanks @fredlf, updated (keeping "whitespace" singular according to previous @jamtur01 review).

**Comment by fredlf** [comment]
Cool, thanks @icecrime. Much appreciated.

File: integration-cli/docker_cli_build_test.go [hunk 4]
```diff
@@ -762,41 +762,188 @@ RUN [ $(ls -l /exists/exists_file | awk '{print $3":"$4}') = 'dockerio:dockerio'
...
+		if _, err := buildImageFromContext(name, ctx, true); err == nil || !strings.Contains(err.Error(), expected) {
+			t.Fatalf("Wrong error: (should contain %q) got:\n%v", expected, err)
+		}
+	}
+
+	for _, addInstruction := range []string{
+		`ADD file1.txt file2.txt test`,
+		`ADD ["file1.txt", "file2.txt", "test"]`,
+	} {
+		testAddFn(t, addInstruction)
```
Comment (LK4D4) [inline integration-cli/docker_cli_build_test.go]: I don't like this, should be separate tests.

**Comment by icecrime** [comment]
@LK4D4 OK: PTAL!

**Comment by icecrime** [comment]
Ping @LK4D4

**Comment by tiborvass** [comment]
LGTM

**Comment by gramster** [comment]
As far as I can tell this doesn't work (Docker 1.9.1):

    Step 23 : ADD [ notebooks/ "/content/Datalab Tutorials and Samples"]
    lstat content/Datalab: no such file or directory

**Comment by EricSalemi** [comment]
4 years have passed... still broken:

Step 9/9 : COPY ['license.clswin01.lic', 'C:\Program Files\MATLAB\R2019a\licenses']
Error response from daemon: failed to process "'C:\\Program": unexpected end of statement while looking for matching single-quote

Is anyone looking at this?
How are we supposed to make this work?

**Comment by thaJeztah** [comment]
@EricSalemi looks like your line has an invalid JSON syntax; JSON does not support single quotes. Try changing to:

```Dockerfile
COPY ["license.clswin01.lic", "C:\Program Files\MATLAB\R2019a\licenses"]
```

**Comment by EricSalemi** [comment]
Same:

    Step 9/9 : COPY ["license.clswin01.lic", "C:\Program Files\MATLAB\R2019a\licenses"]
    Error response from daemon: failed to process "\"C:\\Program":
    unexpected end of statement while looking for matching double-quote

**Comment by EricSalemi** [comment]
I shall mention than I'm using `# escape=` in the beginning of my Dockerfile.

**Comment by thaJeztah** [comment]
Could you open a new ticket with the details requested in the issue template, and a minimal example to reproduce?

**Comment by thaJeztah** [comment]
Oh, might need the backslashes escaped as well inside the JSON (haven't tried, but just in case that's the problem);

```dockerfile
COPY ["license.clswin01.lic", "C:\\Program Files\\MATLAB\\R2019a\\licenses"]
```

**Comment by EricSalemi** [comment]
Yup... that solves it.

Is Docker using a JSON parser here?

**Comment by thaJeztah** [comment]
> Is Docker using a JSON parser here?

Given that these Dockerfile instructions support multiple formats, it:

1.  tries to parse the line as a JSON array.
2. If that fails, it switches to the non-JSON syntax, and treats the line as a `<src> [<src> ...] <dest>`
